dom/events.ts

```diff
@@ -12,17 +12,22 @@ export type EventIntermediateToken = (any : any) => void;
  * @param {Event} event The original event
  * @param {HTMLElement} element The matched element
  */
-export type DelegatedEventHandler = (event : Event, element : HTMLElement) => void;
+export type DelegatedEventHandler<TEvent extends Event = Event, TElement extends HTMLElement = HTMLElement> =
+    (event : TEvent, element : TElement) => void;
 
 type EventHandlerList = {
-    [eventType: string]: EventListener[],
+    [eventType: string]: GenericEventListener<Event>[],
 };
 
+type GenericEventListener<T extends Event = Event> = {
+    (event: T): void;
+}
```

Comment:
Can't we somehow extend the `EventListener` here and make it more generic, so that we don't have to cast everything down there?

Reply:
So you basically want something like this? https://www.typescriptlang.org/play/#src=interface%20GenericEventListener%3CT%20extends%20Event%20%3D%20Event%3E%20extends%20EventListener%20%7B%0D%0A%20%20%20%20(event%3A%20T)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Alet%20a%3A%20GenericEventListener%3CMouseEvent%3E%20%3D%20(event%3A%20MouseEvent)%20%3D%3E%20%7B%20%7D%3B

```typescript
interface GenericEventListener<T extends Event = Event> extends EventListener {
    (event: T): void;
}

let a: GenericEventListener<MouseEvent> = (event: MouseEvent) => { };
```

Reply:
if that works, yes. Because then you can remove all the `as EventListener` further down.

Reply:
I couldn't find a proper way to shadow `EventListener`'s call-signature so that only my generic version is available.

If we extend from `EventListener` the type roughly looks like this:

```typescript
type ResultOfEventListenerAndGenericEventListener<T extends Event> {
    (event: Event): void;
    (event: T): void;
}
```

If I remove the inheritance part, some errors are gone but others occur (due to missing up-casting). As we've concluded via DM, we'll leave it as it is, even though it's inferior as not having the need to cast would be superb. However, if we can't model it with the current type system, we have to roll with whatever works (right now).